libs/@guardian/identity-auth/src/@types/guard.ts

```diff
@@ -0,0 +1,7 @@
+/** A method to create type-guard for string unions */
+export const guard =
+	<T extends readonly unknown[]>(array: T) =>
+	(value: unknown): value is (typeof array)[number] =>
+		array.includes(value);
+
+export type Guard<T> = T extends readonly unknown[] ? T[number] : never;
```

Comment:
I think a version that takes in guard function as argument is harder to misuse.

```suggestion
export type Guard<T extends (value: unknown) => value is unknown> = T extends (
	value: unknown,
) => value is infer G
	? G
	: never;
```

Reply:
when i tried this, typescript was always suggesting it was a `never`

![Screenshot 2023-07-28 at 13 52 50](https://github.com/guardian/csnx/assets/13315440/6b48603a-8c70-44ce-a332-bdfedeffe8fb)

Reply:
linting would also return

```
*/csnx/libs/@guardian/identity-auth/src/cookieRefresh.ts
         15:22  error  Invalid type "never" of template literal expression  @typescript-eslint/restrict-template-expressions
         47:8   error  Unsafe assignment of an `any` value                  @typescript-eslint/no-unsafe-assignment
         47:21  error  Unsafe member access [0] on an `any` value           @typescript-eslint/no-unsafe-member-access
         47:21  error  Unsafe call of an `any` typed value                  @typescript-eslint/no-unsafe-call
         52:38  error  Invalid type "never" of template literal expression  @typescript-eslint/restrict-template-expressions
       
       */csnx/libs/@guardian/identity-auth/src/identityAuth.ts
         52:21  error  Invalid type "never" of template literal expression  @typescript-eslint/restrict-template-expressions
         53:17  error  Invalid type "never" of template literal expression  @typescript-eslint/restrict-template-expressions
         54:16  error  Invalid type "never" of template literal expression  @typescript-eslint/restrict-template-expressions
       
       */csnx/libs/@guardian/identity-auth/src/token.ts
         669:21  error  Unsafe member access [0] on an `any` value           @typescript-eslint/no-unsafe-member-access
         669:21  error  Unsafe call of an `any` typed value                  @typescript-eslint/no-unsafe-call
         745:6   error  Invalid type "never" of template literal expression  @typescript-eslint/restrict-template-expressions
```

Reply:
ah

```
export type ProfileUrl = Guard<typeof isProfileUrl>;
```

Reply:
missed your bit around it taking a guard function

Reply:
Sorry wasn't clear: you're meant to pass your actual type-guard function to it, in this case `isProfileUrl`.